Fix for Incorrect range converting with tilda "~" and "x"

monttda opened the pull request:
Fix of the issue #284 

Change the version number parsing. The changes were done since the new wanted behaviour is that if the version given by the user is of the form '~1.x.x' instead of returning '~> 1.0.0' it returns  '~> 1'. 

This fix does not support version written like "~1.x.3" in this case the output will be inaccurate and will return "~> 1.3"

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/rails-assets/rails-assets/298)

<!-- Reviewable:end -->

Commit:
Change of the Util Module's tests

                Change of 8 tests expected result. This change is related to version number parsing made by the fix_version_string function.
                The changes were done since the new expected behavior will be that when the version given by the user is of the form '~1.x.x'
                instead of returning '~> 1.0.0' it returns  '~> 1'

Commit:
Change on Util module's functionality

                Change to the fix_version_string function that does the version number parsing .
                The changes were done since the new wanted behavior is that if version given by the user is of the form '~1.x.x'
                instead of returning '~> 1.0.0' it returns  '~> 1'

jandudulski commented:
> This fix does not support version written like "~1.x.3" in this case the output will be inaccurate and will return "~> 1.3"

@monttda can you prepare fix so it would transform `~1.x.3` into `~> 1.0.3`?

Commit:
Replacement of the middle number for a 0

  In the case that the middle number is a x (ex. 1.x.4)
  It is now replaced by a 0 insted of removing the .x.

  A test was also added to verify this case.

jandudulski commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

_[spec/models/build/utils_spec.rb, line 51 [r2]](https://reviewable.io:443/reviews/rails-assets/rails-assets/298#-K8JkiCUTH1ex5oDkkmu:-K8JkiCVGLKFAgy9P8p7:1299446410) ([raw file](https://github.com/rails-assets/rails-assets/blob/169a1fbe91a2a662cdbc8b2f338f0912951ecdf9/spec/models/build/utils_spec.rb#L51)):_
Actually this doesn't seems to be correct. 

`~> 2.0.3` matches `>= 2.0.3` and `< 2.1.0` but
`2.x.3` should (I guess) match `2.0.3`, `2.1.3`, `2.2.3`, ... so something completely not supported by bundler.

@sheerun could you please confirm how it works in bower for `x` in the middle?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/rails-assets/rails-assets/298)_

<!-- Sent from Reviewable.io -->

sheerun commented:
2.x.3 seems the same as 2.x.x, and it's equivalent of ^2.0 in npm and ~> 2.0 in bundler

This PR makes following wrong changes:
- `2.x.3` to mean `~> 2.0.3` (it means `~> 2.0`)
- `~1.x`, `1.x`, `~1.*` and `1.*` to mean `~> 1` (it means `~> 1.0`)
- `~v1.0.x` to mean `~> 1.0` (it means `~> 1.0.0`)

jandudulski commented:
Thanks Adam for clarification!
